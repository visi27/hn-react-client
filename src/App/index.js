import { connect } from 'react-redux';
import { fetchData } from '../_actions/result';
import App from './presenter';

const mapDispatchToProps = dispatch => ({
  onLoad() {
    dispatch(fetchData({}));
  },
  onLocationChange() {
    dispatch(fetchData({}));
  },
});

export default connect(null, mapDispatchToProps)(App);
