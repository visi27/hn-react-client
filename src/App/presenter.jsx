import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { Route } from 'react-router-dom';
import Nav from '../Nav/index';
import ResultsView from '../ResultsView/index';
import './style.css';

class App extends Component {
  componentDidMount() {
    this.props.onLoad();
  }
  shouldComponentUpdate(nextProps) {
    return this.props.location.pathname !== nextProps.location.pathname;
  }

  componentDidUpdate() {
    this.props.onLocationChange();
  }

  render() {
    return (
      <div>
        <Route path="/" component={Nav} />
        <main role="main" className="mt-3">
          <ResultsView />
        </main>
      </div>
    );
  }
}

App.propTypes = {
  onLoad: PropTypes.func.isRequired,
  onLocationChange: PropTypes.func.isRequired,
  location: PropTypes.shape({
    pathname: PropTypes.string,
  }).isRequired,
};

export default App;
