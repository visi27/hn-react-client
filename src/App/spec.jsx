import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import React from 'react';
import App from './presenter';

configure({ adapter: new Adapter() });
const props = {
  onLoad: () => {},
  onLocationChange: () => {},
  location: { pathname: '/' },
};
describe('App', () => {
  test('has a valid snapshot', () => {
    const component = shallow(<App {...props} />);
    const tree = toJson(component);
    expect(tree).toMatchSnapshot();
  });
});
