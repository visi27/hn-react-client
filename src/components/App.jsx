import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import Search from './Search/index';
import Table from './Table/index';
import './App.css';
import Button from './Button/presenter';
import withLoading from '../hoc/withLoading';
import fetchResults from '../actions/result';
import loadMenu from '../actions/menu';
import Header from './Header/presenter';

const ButtonWithLoading = withLoading(Button);

export class App extends Component {
  componentDidMount() {
    this.props.onLoad();
  }

  render() {
    return (
      <div className="page">
        <Header />
        <span>HELLO</span>
        <div className="interactions">
          <Search> Search </Search>
        </div>

        <Table />

        <ButtonWithLoading>More Articles</ButtonWithLoading>
      </div>
    );
  }
}

App.propTypes = {
  onLoad: PropTypes.func.isRequired,
};

function mapDispatchToProps(dispatch) {
  return {
    onLoad: () => {
      dispatch(loadMenu());
      dispatch(fetchResults());
    },
  };
}

export default connect(null, mapDispatchToProps)(App);
