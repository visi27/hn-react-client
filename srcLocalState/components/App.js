// @flow

import React, { Component } from 'react';
import Search from './Search/';
import Table from './Table/index';
import './App.css';
import Button from './Button/';
import withLoading from '../hoc/withLoading';

import { AppConfig as conf } from '../config';

const ButtonWithLoading = withLoading(Button);

type Props = {
  foo: number
};

type Result = {
  hits: Array<Object>,
  page: number,
  nbPages: number
};

type State = {
  searchTerm: string,
  currentPage: number,
  result: Result,
  cache: Object,
  error: ?Object,
  isLoading: boolean,
};

class App extends Component<Props, State> {
  setSearchTopStories: (result: Result, updateCache: boolean) => void;
  fetchSearchTopStories: (searchTerm: string, currentPage: number) => void;
  onDismiss: () => void;
  onSearchChange: () => void;
  onSearchSubmit: () => void;
  onReset: () => void;
  nextPage: () => void;

  constructor (props: Props) {
    super(props);
    this.state = {
      searchTerm: conf.DEFAULT_QUERY,
      currentPage: 0,
      cache: {},
      result: {
        hits: [],
        page: 0,
        nbPages: 0,
      },
      error: null,
      isLoading: false,
    };

    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
    this.onReset = this.onReset.bind(this);
    this.nextPage = this.nextPage.bind(this);
  }

  componentDidMount () {
    const {searchTerm, currentPage} = this.state;
    this.fetchSearchTopStories(searchTerm, currentPage);
  }

  render () {
    const {searchTerm, result, currentPage, error, isLoading} = this.state;
    const totPages = result ? result.nbPages : 0;

    return (
      <div className="page">
        <span>HELLO</span>
        <div className="interactions">
          <Search
            searchTerm={searchTerm}
            onSearchChange={this.onSearchChange}
            onSearchSubmit={this.onSearchSubmit}
            onReset={this.onReset}
          >
            Search
          </Search>
        </div>
        {error ? (
          <div className="interactions">
            <p>Application Error: {error.message}</p>
          </div>
        ) : (
          result.hits.length > 0 && (
            <Table
              list={result.hits}
              searchTerm={searchTerm}
              onDismiss={this.onDismiss}
            />
          )
        )}

        {currentPage < totPages && (
          <ButtonWithLoading onClick={this.nextPage} isLoading={isLoading}>
            More Articles
          </ButtonWithLoading>
        )}

      </div>
    );
  }

  nextPage () {
    const {searchTerm, currentPage, result} = this.state;
    const totPages = result.nbPages;

    if (currentPage < totPages) {
      this.setState({
        currentPage: currentPage + 1,
      });
    }

    this.fetchSearchTopStories(searchTerm, currentPage + 1);
  }

  setSearchTopStories (result: Result, updateCache: boolean) {
    const {hits, page, nbPages} = result;

    this.setState(updateSearchTopStoriesState(hits, page, nbPages, updateCache));
  }

  fetchSearchTopStories (searchTerm: string, currentPage: number) {
    this.setState({isLoading: true});
    const {cache} = this.state;

    cache[searchTerm] && cache[searchTerm]['pages'][currentPage]
      ? this.setSearchTopStories(cache[searchTerm]['pages'][currentPage], false)
      : fetch(
      `${conf.PATH_BASE}${conf.PATH_SEARCH}?${conf.PARAM_SEARCH}${searchTerm}&${conf.PARAM_PAGE}${currentPage}&${conf.PARAM_HPP}${conf.DEFAULT_HPP}`,
      )
        .then(response => response.json())
        .then(result => this.setSearchTopStories(result, true))
        .catch(e => this.setState({error: e}));
  }

  onDismiss (itemId: number) {
    const updatedHits = this.state.result.hits.filter(item => {
      return item.objectID !== itemId;
    });

    this.setState({
      result: {...this.state.result, hits: updatedHits},
    });
  }

  onSearchChange (event: Object) {
    this.setState({searchTerm: event.target.value});
  }

  onSearchSubmit (event: Object) {
    this.setState({currentPage: 0});
    const {searchTerm} = this.state;
    this.fetchSearchTopStories(searchTerm, 0);
    event.preventDefault();
  }

  onReset () {
    this.setState({
      result: {
        hits: [],
        page: 0,
        nbPages: 0,
      },
      searchTerm: '',
    });
  }
}

const updateSearchTopStoriesState = (hits, page, nbPages, updateCache) => (prevState: State) => {

  const oldHits = page !== 0 ? prevState.result.hits : [];
  const updatedHits = [...oldHits, ...hits];

  let {cache} = prevState;
  if (updateCache) {
    if (!cache[prevState.searchTerm]) {
      cache[prevState.searchTerm] = {pages: []};
    }
    cache[prevState.searchTerm]['pages'][page] = {hits, page, nbPages};
  }

  return {
    result: {hits: updatedHits, page, nbPages},
    cache: cache,
    isLoading: false,
  };
};

export default App;
