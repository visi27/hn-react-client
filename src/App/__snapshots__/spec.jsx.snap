// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `
<div>
  <Route
    component={[Function]}
    path="/"
  />
  <main
    className="mt-3"
    role="main"
  >
    <Connect(Table) />
  </main>
</div>
`;
